L74-java-channel-creds.md

```diff
@@ -0,0 +1,485 @@
+
+  /* Other methods can be added to support various Features */
+
+  public enum Feature {}
+
+  public final static class Builder {
+    public TlsChannelCredentials build() {...}
+  }
+}
+```
+
+An example `Feature` could be `CLIENT_CERTIFICATE`. When the `Feature` is
+added, methods `getCertificateChain()`, `getPrivateKey()`, `getPassword()` can
+be added. Observing the contents of those methods would be a requirement for
+`CLIENT_CERTIFICATE`. An implementation understanding `CLIENT_CERTIFICATE`
+might not support encrypted private keys and so could consider the credential
+unsupported if `getPassword()` returned non-`null`.
+
+To support very advanced use cases, Netty will provide a credential that wraps
+a `ProtocolNegotiator`. This allows implementations like ALTS and XDS to use
+internal APIs without forcing their users to use experimental or internal APIs,
+as their users would just interact with `ChannelCredentials`.
+
+```java
+package io.grpc.netty;
+
+final class NettyChannelCredentials extends ChannelCredentials {
+  public NettyChannelCredentials(ProtocolNegotiator.Factory negotiator) {...}
+  public ProtocolNegotiator.Factory getNegotiator() { return negotiator; }
+}
+
+@Internal
+public final class InternalNettyChannelCredentials {
+  private InternalNettyChannelCredentials() {}
+
+  public ChannelCredentials create(InternalProtocolNegotiator.Factory negotiator)
+    {...}
+
+  /**
+   * Converts credentials to a negotiator, in similar fashion as for a new
+   * channel.
+   *
+   * @throws IllegalArgumentException if unable to convert
+   */
+  public InternalProtocolNegotiator.Factory toNegotiator(ChannelCredentials creds)
+    {...}
+}
+```
+
+### GoogleDefault ChannelCredentials Example
+
+The API provided here is not part of the design. It is meant as a
+demonstration.
+
+```java
+public final class GoogleDefaultChannelCredentials {
+  private GoogleDefaultChannelCredentials() {}
+
+  public static ChannelCredentials create() {
+    return new ChoiceChannelCredentials(
+        InternalNettyChannelCredentials.create(
+          new AltsProtocolNegotiatorFactory()),
+        new CompositeChannelCredentials(
+          new TlsChannelCredentials(),
+          MoreCallCredentials.from(GoogleCredentials.getApplicationDefault())));
+  }
+}
+```
+
+### xDS ChannelCredentials Example
+
+The API provided here is not part of the design. It is meant as a
+demonstration.
+
+```java
+public final class XdsChannelCredentials {
+  private XdsChannelCredentials() {}
+
+  /**
+   * Creates credentials to be configured by xDS, falling back to other
+   * credentials if no configuration is provided.
+   *
+   * @throws IllegalArgumentException if fallback is unable to be used
+   */
+  public static ChannelCredentials createWithFallback(
+      ChannelCredentials fallback) {
+    return new InternalNettyChannelCredentials.create(
+        new XdsProtocolNegotiatorFactory(
```

Comment:
Is this meant to be the existing class [SdsProtocolNegotiators.ClientSdsProtocolNegotiatorFactory](https://github.com/grpc/grpc-java/blob/master/xds/src/main/java/io/grpc/xds/internal/sds/SdsProtocolNegotiators.java#L81) - renamed for convenience? Currently it doesn't accept a fallback-negotiator (but that's to be fixed there and not in this doc)

Reply:
It very well could be. If SdsProtocolNegotiators _only_ supports SDS, then maybe there will be some other negotiator that will wrap it and allow using SDS or Mesh CA or some other such things. The `XdsProtocolNegotiatorFactory` name here doesn't matter as it should be package-private/internal; name it whatever you want.

Reply:
`SdsProtocolNegotiators` is currently misnamed (like many other classes in the package) because it's not tied to SDS but rather to xDS-provided security info so it supports all possible credential-types provided by xDS (file-based, SDS based and the new cert-provider plugin based).